pkg/server/cloud/aws/instances.go

```diff
@@ -647,12 +662,59 @@ func isUnsupportedInstanceError(err error) bool {
 // InvalidAvailabilityZone
 
 func (e *AwsEC2) AddIAMPermissions(node *api.Node, instanceProfile string) error {
-	_, err := e.client.AssociateIamInstanceProfile(
-		&ec2.AssociateIamInstanceProfileInput{
-			IamInstanceProfile: &ec2.IamInstanceProfileSpecification{
-				Arn: aws.String(instanceProfile),
+	instanceID := node.Status.InstanceID
+	out, err := e.client.DescribeIamInstanceProfileAssociations(
+		&ec2.DescribeIamInstanceProfileAssociationsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("instance-id"),
+					Values: aws.StringSlice([]string{instanceID}),
+				},
 			},
-			InstanceId: aws.String(node.Status.InstanceID),
 		})
-	return err
+	if err != nil {
+		return util.WrapError(err, "DescribeIamInstanceProfileAssociations() "+instanceID)
+	}
+
+	if out == nil || len(out.IamInstanceProfileAssociations) == 0 {
+		klog.V(5).Infof("adding IAM instance profile %s to instance %s", instanceProfile, instanceID)
+		_, err = e.client.AssociateIamInstanceProfile(
+			&ec2.AssociateIamInstanceProfileInput{
+				IamInstanceProfile: getIAMInstanceProfileSpecification(instanceProfile),
+				InstanceId:         aws.String(instanceID),
+			})
+		if err != nil {
+			return util.WrapError(err, "AssociateIamInstanceProfile() "+instanceID)
+		}
+		klog.V(5).Infof("added IAM instance profile %s to %s", instanceProfile, instanceID)
+		return nil
+	}
+
+	n := len(out.IamInstanceProfileAssociations)
+	if n > 1 {
+		klog.Warningf("DescribeIamInstanceProfileAssociations() %s: got %d results", instanceID, n)
+	}
+
+	for _, association := range out.IamInstanceProfileAssociations {
+		if association == nil {
+			return fmt.Errorf("DescribeIamInstanceProfileAssociations() %s: nil association", instanceID)
+		}
+		klog.V(5).Infof("replacing current IAM instance profile %s on %s association ID %s state %s",
+			aws.StringValue(out.IamInstanceProfileAssociations[0].IamInstanceProfile.Arn),
```

Comment:
Are those log lines correct? Shouldn't we log
`aws.StringValue(association.IamInstanceProfile.Arn)` instead of always logging first element from `out.IamInstanceProfileAssociations`?

Reply:
Yeah, I added a `break` at the end of the loop. The API only returns one result, but better be making sure that we're not trying to update the association more than once.